crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,801 @@
+///     
+///     complex double coeffs[2] = {1, -1};
+///     QkBitTerm bits[4] = {QkBitTerm_Zero, QkBitTerm_One, QkBitTerm_Plus, QkBitTerm_Minus};
+///     uint32_t indices[4] = {0, 1, 98, 99};  // <-- e.g. {1, 0, 99, 98} would be invalid
+///     size_t boundaries[3] = {0, 2, 4};
+///     
+///     QkSparseObservable *obs = qk_obs_new(
+///         num_qubits, num_terms, num_bits, coeffs, bits, indices, boundaries
+///     );
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following conditions are violated:
+///
+///   * ``coeffs`` is a pointer to a ``complex double`` array of length ``num_terms``
+///   * ``bit_terms`` is a pointer to a ``QkBitTerm`` array of length ``num_bits``
+///   * ``indices`` is a pointer to a ``uint32_t`` array of length ``num_bits``, which is
+///     term-wise sorted in strict ascending order, and every element is smaller than ``num_qubits``
+///   * ``boundaries`` is a pointer to a ``size_t`` array of length ``num_terms + 1``, which is
+///     sorted in ascending order, the first element is 0 and the last element is
+///     smaller than ``num_terms``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_new(
+    num_qubits: u32,
+    num_terms: u64,
+    num_bits: u64,
+    coeffs: *mut Complex64,
+    bit_terms: *mut BitTerm,
+    indices: *mut u32,
+    boundaries: *mut usize,
+) -> *mut SparseObservable {
+    let num_terms = num_terms as usize;
+    let num_bits = num_bits as usize;
+
+    check_ptr(coeffs).unwrap();
+    check_ptr(bit_terms).unwrap();
+    check_ptr(indices).unwrap();
+    check_ptr(boundaries).unwrap();
+
+    // SAFETY: At this point we know the pointers are non-null and aligned. We rely on C that
+    // the pointers point to arrays of appropriate length, as specified in the function docs.
+    let coeffs = unsafe { ::std::slice::from_raw_parts(coeffs, num_terms).to_vec() };
+    let bit_terms = unsafe { ::std::slice::from_raw_parts(bit_terms, num_bits).to_vec() };
+    let indices = unsafe { ::std::slice::from_raw_parts(indices, num_bits).to_vec() };
+    let boundaries = unsafe { ::std::slice::from_raw_parts(boundaries, num_terms + 1).to_vec() };
+
+    let result = SparseObservable::new(num_qubits, coeffs, bit_terms, indices, boundaries);
+    match result {
+        Ok(obs) => Box::into_raw(Box::new(obs)),
+        Err(_) => ::std::ptr::null_mut(),
+    }
+}
+
+/// @ingroup QkSparseObservable
+/// Free the observable.
+///
+/// @param obs A pointer to the observable to free.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_zero(100);
+///     qk_obs_free(obs);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``obs`` is not either null or a valid pointer to a
+/// [SparseObservable].
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_free(obs: *mut SparseObservable) {
+    if !obs.is_null() {
+        if !obs.is_aligned() {
+            panic!("Attempted to free a non-aligned pointer.")
```

Comment:
Should we have a panic if it's a `null` pointer too with an error message that says as much?

Reply:
I didn't add this since it seems valid C code to free null pointers (so nothing happens), but we can also add it for code validation

Reply:
I just don't like leaving it undefined behavior. I'd rather just explicitly error even if C doesn't care.

Reply:
`free(NULL)` is well defined to do nothing in C.  It's not so much "doesn't care" as "explicitly not an error".